go/src/koding/klient/machine/transport/rsync/rsync.go

```diff
@@ -0,0 +1,204 @@
+	// current status of downloading will be periodically sent to provided
+	// progress callback function. io.EOF error is sent to the callback when
+	// downloading is complete.
+	Progress func(n, size, speed int64, err error)
+}
+
+// valid checks if command fields are valid.
+func (c *Command) valid() error {
+	if c == nil {
+		return errors.New("rsync: command is nil")
+	}
+
+	if c.SourcePath == "" {
+		return errors.New("rsync: source path is not set")
+	}
+	if c.DestinationPath == "" {
+		return errors.New("rsync: destination path is not set")
+	}
+
+	return nil
+}
+
+// Run starts new rsync process. And waits for it to complete.
+func (c *Command) Run(ctx context.Context) error {
+	if err := c.valid(); err != nil {
+		return err
+	}
+
+	if c.Cmd == nil {
+		c.Cmd = exec.CommandContext(ctx, "rsync")
+		c.Cmd.Env = os.Environ()
+	}
+
+	// Add default arguments.
+	c.Cmd.Args = append(c.Cmd.Args, "-zlptgoDd")
+
+	// Use remote shell if SSH private key path is set.
+	if c.PrivateKeyPath != "" {
+		// TODO(ppknap): check if RC4 cipher will work on every machine without
+		// altering sshd_config on destination.
+		rsh := []string{
+			"ssh", "-T", "-x", "-i", c.PrivateKeyPath,
+			"-oCompression=no",
+			"-oStrictHostKeychecking=no",
+		}
+
+		if c.SSHPort > 0 {
+			rsh = append(rsh, " -p ", strconv.Itoa(c.SSHPort))
+		}
+
+		c.Cmd.Args = append(c.Cmd.Args, "-e", strings.Join(rsh, " "))
+	}
+
+	// Progress logic needs verbose mode with itemized changes.
+	if c.Progress != nil {
+		c.Cmd.Args = append(c.Cmd.Args, "-Priv")
+	}
+
+	if c.Username != "" && c.Host != "" {
+		if c.Download {
+			c.SourcePath = c.Username + "@" + c.Host + ":" + c.SourcePath
+		} else {
+			c.DestinationPath = c.Username + "@" + c.Host + ":" + c.DestinationPath
+		}
+	}
+	c.Cmd.Args = append(c.Cmd.Args, c.SourcePath, c.DestinationPath)
+
+	if c.Progress == nil {
+		return c.Cmd.Run()
+	}
+
+	// Set up progress callback when it's provided.
+	rc, err := c.Cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+	defer rc.Close()
+
+	if err := c.Cmd.Start(); err != nil {
+		return err
+	}
+
+	c.scan(rc)
+	return c.Cmd.Wait()
+}
+
+var (
+	bitRe  = regexp.MustCompile(`^[.><ch*].......... .`)
+	sizeRe = regexp.MustCompile(`^\s*([\d,]+)\s+\d+%.*$`)
+)
+
+func (c *Command) scan(r io.Reader) {
+	defer c.Progress(0, 0, 0, io.EOF)
+
+	var (
+		now           time.Time
+		n, size, part int64
+	)
+
+	scanner := bufio.NewScanner(r)
+	scanner.Split(scanNonControl)
+	for scanner.Scan() {
+		if now.IsZero() {
+			now = time.Now()
+		}
+
+		line := scanner.Text()
+		if bitRe.MatchString(line) {
+			size += part
+			n++
+			part = 0
+			continue
+		}
+
+		ms := sizeRe.FindStringSubmatch(line)
+		if len(ms) < 2 {
+			continue
+		}
+
+		p, err := strconv.Atoi(strings.Replace(ms[1], ",", "", -1))
+		if err != nil {
+			continue
+		}
+
+		part = int64(p)
+
+		speed := int64(float64(part+size) / float64(time.Since(now)/time.Second))
```

Comment:
Wouldn't you want to take ceiling here, to have `speed := 3` when `speed := int64(2.9)`?

```
speed := int64(float64(part+size) / float64(time.Since(now)/time.Second) + 0.5)
```